cpp/include/raft/sparse/hierarchy/detail/connectivities.cuh

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/cudart_utils.h>
+#include <raft/cuda_utils.cuh>
+#include <raft/handle.hpp>
+
+#include <raft/linalg/unary_op.cuh>
+
+#include <raft/linalg/distance_type.h>
+#include <raft/sparse/hierarchy/common.h>
+#include <raft/mr/device/buffer.hpp>
+#include <raft/sparse/convert/csr.cuh>
+#include <raft/sparse/coo.cuh>
+#include <raft/sparse/selection/knn_graph.cuh>
+
+#include <limits>
+
+namespace raft {
+namespace hierarchy {
+namespace detail {
+
+template <raft::hierarchy::LinkageDistance dist_type, typename value_idx,
+          typename value_t>
+struct distance_graph_impl {
+  void run(const raft::handle_t &handle, const value_t *X, size_t m, size_t n,
+           raft::distance::DistanceType metric,
+           raft::mr::device::buffer<value_idx> &indptr,
+           raft::mr::device::buffer<value_idx> &indices,
+           raft::mr::device::buffer<value_t> &data, int c);
+};
+
+/**
+ * Connectivities specialization to build a knn graph
+ * @tparam value_idx
+ * @tparam value_t
+ */
+template <typename value_idx, typename value_t>
+struct distance_graph_impl<LinkageDistance::KNN_GRAPH, value_idx, value_t> {
```

Comment:
Is `LinkageDistance::Pairwise` not supported yet? Also fully qualified namespaces to resolve ambiguity when namespace is not a direct child

Reply:
The `Pairwise` option lives in cuML at the moment while we await the removal of cutlass from our dense pairwise distances API.